cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -398,6 +403,110 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 	return nil
 }
 
+func (a *StaticAutoscaler) handleQuotaAndStockoutErrors(currentTime time.Time) {
+	currentInstancesPerNodeGroup := a.clusterStateRegistry.GetCloudProviderNodeInstances()
+	previousInstancesPerNodeGroup := a.clusterStateRegistry.GetPreviousCloudProviderNodeInstances()
+	nodeGroups := a.CloudProvider.NodeGroups()
+
+	for _, nodeGroup := range nodeGroups {
+		a.handleQuotaAndStockoutErrorsForNodeGroup(
+			nodeGroup,
+			currentInstancesPerNodeGroup[nodeGroup.Id()],
+			previousInstancesPerNodeGroup[nodeGroup.Id()],
+			currentTime)
+	}
+}
+
+func (a *StaticAutoscaler) handleQuotaAndStockoutErrorsForNodeGroup(
+	nodeGroup cloudprovider.NodeGroup,
+	currentInstances []cloudprovider.Instance,
+	previousInstances []cloudprovider.Instance,
+	currentTime time.Time) {
+
+	_, currentErrorCodeToInstance := a.buildInstanceToOutOfResourcesErrorCodeMappings(currentInstances)
+	previousInstanceToErrorCode, _ := a.buildInstanceToOutOfResourcesErrorCodeMappings(previousInstances)
+
+	// If node group is scaling up and there are new node-create requests which cannot be satisfied because of
+	// out-of-resources errors we:
+	//  - emit event
+	//  - alter the scale-up
+	//  - increase scale-up failure metric
+	//  - backoff the node group
+	for errorCode, instanceIds := range currentErrorCodeToInstance {
+		unseenInstanceIds := make([]string, 0)
+		for _, instanceId := range instanceIds {
+			if _, seen := previousInstanceToErrorCode[instanceId]; !seen {
+				unseenInstanceIds = append(unseenInstanceIds, instanceId)
+			}
+		}
+
+		klog.Infof("Failed adding %v nodes (%v unseen previously) to group %v due to %v", len(instanceIds), len(unseenInstanceIds), nodeGroup.Id(), errorCode)
+		if len(unseenInstanceIds) > 0 && a.clusterStateRegistry.IsNodeGroupScalingUp(nodeGroup.Id()) {
+			a.LogRecorder.Eventf(
+				apiv1.EventTypeWarning,
+				"ScaleUpFailed",
+				"Failed adding %v nodes to group %v due to %v",
```

Comment:
Cloudprovider gives us ErrorMessage, it would be nice to expose it to user (otherwise what's the point of having it?).

Reply:
Added log as we discussed.